src/OrleansAWSUtils/Reminders/DynamoDBReminderTable.cs

```diff
@@ -0,0 +1,294 @@
+                var records = await storage.ScanAsync(TABLE_NAME_DEFAULT_VALUE, expressionValues, expression,
+                    item => new Dictionary<string, AttributeValue>
+                    {
+                        { REMINDER_ID_PROPERTY_NAME, item[REMINDER_ID_PROPERTY_NAME] },
+                        { GRAIN_HASH_PROPERTY_NAME, item[GRAIN_HASH_PROPERTY_NAME] }
+                    }).ConfigureAwait(false);
+
+                if (records.Count <= 25)
+                {
+                    await storage.DeleteEntriesAsync(TABLE_NAME_DEFAULT_VALUE, records);
+                }
+                else
+                {
+                    List<Task> tasks = new List<Task>();
+                    foreach (var batch in records.BatchIEnumerable(25))
+                    {
+                        tasks.Add(storage.DeleteEntriesAsync(TABLE_NAME_DEFAULT_VALUE, batch));
+                    }
+                    await Task.WhenAll(tasks);
+                }
+            }
+            catch (Exception exc)
+            {
+                logger.Warn(ErrorCode.ReminderServiceBase,
+                    $"Intermediate error removing reminder entries {Utils.DictionaryToString(expressionValues)} from table {TABLE_NAME_DEFAULT_VALUE}.", exc);
+                throw;
+            }
+        }
+
+        public async Task<string> UpsertRow(ReminderEntry entry)
+        {
+            var reminderId = ConstructReminderId(serviceId, entry.GrainRef, entry.ReminderName);
+            var keys = new Dictionary<string, AttributeValue>
+                {
+                    { REMINDER_ID_PROPERTY_NAME, new AttributeValue(reminderId) },
+                    { GRAIN_HASH_PROPERTY_NAME, new AttributeValue { N = entry.GrainRef.GetUniformHashCode().ToString() } }
+                };
+
+            var fields = new Dictionary<string, AttributeValue>
+                {
+                    { SERVICE_ID_PROPERTY_NAME, new AttributeValue(serviceId.ToString()) },
+                    { GRAIN_REFERENCE_PROPERTY_NAME, new AttributeValue( entry.GrainRef.ToKeyString()) },
+                    { PERIOD_PROPERTY_NAME, new AttributeValue(entry.Period.ToString()) },
+                    { START_TIME_PROPERTY_NAME, new AttributeValue(entry.StartAt.ToString()) },
+                    { REMINDER_NAME_PROPERTY_NAME, new AttributeValue(entry.ReminderName) }
+                };
+
+            var extraExpression = $"{ETAG_PROPERTY_NAME} = {ETAG_PROPERTY_NAME} + {ETAG_INCREMENT}";
+            var etagField = new Dictionary<string, AttributeValue>
+                {
+                    { ETAG_INCREMENT, new AttributeValue { N = "1" } }
+                };
+
+            try
+            {
+                if (logger.IsVerbose) logger.Verbose("UpdateRow entry = {0}, etag = {1}", entry.ToString(), entry.ETag);
+
+                var foundETag = await storage.ReadSingleEntryAsync(TABLE_NAME_DEFAULT_VALUE, keys, item =>
```

Comment:
The parallel upsert test should succeed: reminders have different names there. Name is part of row key.